File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/infrastructure/PostgresValorantStoreHistoryRepository.java
package fr.huiitre.tools.modules.riot.valorant.infrastructure;

import fr.huiitre.tools.modules.riot.valorant.application.user.ports.ValorantStoreHistoryRepository;
import org.springframework.jdbc.core.JdbcTemplate;

import java.time.LocalDate;
import java.util.*;

public class PostgresValorantStoreHistoryRepository implements ValorantStoreHistoryRepository {

    private final JdbcTemplate jdbcTemplate;

    public PostgresValorantStoreHistoryRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    @Override
    public Map<LocalDate, List<Long>> findAllRawByAccountId(Long accountId) {
        final String sql = """
                SELECT seen_at, skin_id
                FROM tools_riot.valorant_store_history
                WHERE valorant_account_id = ?
                ORDER BY seen_at DESC
                """;

        Map<LocalDate, List<Long>> history = new LinkedHashMap<>();
        jdbcTemplate.query(sql, rs -> {
            LocalDate date = rs.getDate("seen_at").toLocalDate();
            Long skinId = rs.getLong("skin_id");
            history.computeIfAbsent(date, k -> new ArrayList<>()).add(skinId);
        }, accountId);

        return history;
    }

    @Override
    public Long add(Long accountId, Long skinId, LocalDate seenAt) {
        final String sql = "INSERT INTO tools_riot.valorant_store_history (valorant_account_id, skin_id, seen_at) VALUES (?, ?, ?) RETURNING id";
        return jdbcTemplate.queryForObject(sql, Long.class, accountId, skinId, seenAt);
    }

    @Override
    public boolean existsByAccountIdAndSkinIdAndDate(Long accountId, Long skinId, LocalDate seenAt) {
        final String sql = "SELECT EXISTS (SELECT 1 FROM tools_riot.valorant_store_history WHERE valorant_account_id = ? AND skin_id = ? AND seen_at = ?)";
        return Boolean.TRUE.equals(jdbcTemplate.queryForObject(sql, Boolean.class, accountId, skinId, seenAt));
    }
}

File: api/Modules/Riot/Valorant/Infrastructure/PostgresValorantStoreHistoryRepository.cs
using Tools.Api.Modules.Riot.Common.Infrastructure;
using Tools.Api.Modules.Riot.Valorant.Application.User.Ports;

namespace Tools.Api.Modules.Riot.Valorant.Infrastructure;

public sealed class PostgresValorantStoreHistoryRepository(RiotDatabase database) : IValorantStoreHistoryRepository
{
    private sealed record HistoryRow(DateOnly SeenAt, long SkinId);

    public async Task<Dictionary<DateOnly, List<long>>> FindAllRawByAccountId(long accountId)
    {
        var rows = await database.Query<HistoryRow>(
            """
            SELECT seen_at AS SeenAt, skin_id AS SkinId
            FROM tools_riot.valorant_store_history
            WHERE valorant_account_id = @AccountId
            ORDER BY seen_at DESC
            """,
            new { AccountId = accountId });

        // Le tri est porté par la requête : le regroupement conserve l'ordre des lignes, donc les
        // journées sortent de la plus récente à la plus ancienne.
        return rows
            .GroupBy(row => row.SeenAt)
            .ToDictionary(group => group.Key, group => group.Select(row => row.SkinId).ToList());
    }

    public Task<long> Add(long accountId, long skinId, DateOnly seenAt) =>
        database.ExecuteScalar<long>(
            """
            INSERT INTO tools_riot.valorant_store_history (valorant_account_id, skin_id, seen_at)
            VALUES (@AccountId, @SkinId, @SeenAt)
            RETURNING id
            """,
            new { AccountId = accountId, SkinId = skinId, SeenAt = seenAt });

    public Task<bool> ExistsByAccountIdAndSkinIdAndDate(long accountId, long skinId, DateOnly seenAt) =>
        database.ExecuteScalar<bool>(
            """
            SELECT EXISTS (
                SELECT 1 FROM tools_riot.valorant_store_history
                WHERE valorant_account_id = @AccountId AND skin_id = @SkinId AND seen_at = @SeenAt
            )
            """,
            new { AccountId = accountId, SkinId = skinId, SeenAt = seenAt });
}
